tensorzero-core/src/experimentation/mod.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::BTreeMap;
+use std::{collections::HashMap, sync::Arc};
+
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use uuid::Uuid;
+
+use crate::error::{Error, ErrorDetails, IMPOSSIBLE_ERROR_MESSAGE};
+use crate::variant::VariantInfo;
+
+mod static_weights;
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[cfg_attr(test, derive(ts_rs::TS))]
+#[cfg_attr(test, ts(export))]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ExperimentationConfig {
+    StaticWeights(static_weights::StaticWeightsConfig),
+    #[default]
+    Uniform,
+}
+
+pub trait VariantSampler {
+    // TODO, when we add bandits: pass CH and PG clients here (but use opaque trait types)
+    async fn setup(&self) -> Result<(), Error>;
+    async fn sample(
+        &self,
+        function_name: &str,
+        episode_id: Uuid,
+        active_variants: &mut BTreeMap<String, Arc<VariantInfo>>,
+    ) -> Result<(String, Arc<VariantInfo>), Error>;
+}
+
+impl ExperimentationConfig {
+    /// Note: in the future when we deprecate variant.weight we can simply use #[serde(default)]
+    /// and default to ExperimentationConfig::Uniform
+    ///
+    /// For now, we call this function from the
+    pub fn legacy_from_variants_map(variants: &HashMap<String, Arc<VariantInfo>>) -> Self {
+        // We loop over the variants map and if any of them have weights we output a StaticWeightsConfig
+        // otherwise, we output a Uniform config
+        for variant in variants.values() {
+            if variant.inner.weight().is_some() {
+                return Self::StaticWeights(
+                    static_weights::StaticWeightsConfig::legacy_from_variants_map(variants),
+                );
+            }
+        }
+        Self::Uniform
+    }
+}
+
+impl VariantSampler for ExperimentationConfig {
+    async fn setup(&self) -> Result<(), Error> {
+        match self {
+            Self::StaticWeights(config) => config.setup().await,
+            Self::Uniform => Ok(()),
+        }
+    }
+
+    async fn sample(
+        &self,
+        function_name: &str,
+        episode_id: Uuid,
+        active_variants: &mut BTreeMap<String, Arc<VariantInfo>>,
+    ) -> Result<(String, Arc<VariantInfo>), Error> {
+        match active_variants.len() {
+            0 => Err(Error::new(ErrorDetails::Inference {
+                message: format!(
+                    "VariantSampler::sample called with no active variants. {IMPOSSIBLE_ERROR_MESSAGE}"
+                ),
+            })),
+            1 => {
```

Comment:
Can you add a comment here, noting that we'll probably still want to check-and-set this variant in Postgres when we add bandits? This will allow something like `inference(episode_id=abc) -> <gateway shutdown> -> <add variants> -> inference(episode_id=abc)` to use the same variant.